### ISSUE TITLE
Question: Multiple OEs within certificate

### ISSUE THREAD

**Issue by AlexThurston** [usnistgov/ACVP#799]
This is more of a question than a bug.

It has to do with the organization of certificates/modules/oes.  I gather that it is possible (although it does not seem to be the norm) that a single module (or product as it's called on the CAVP list) can have multiple certificates.  Trolling through the current certificate list there are a handful of examples.  However, is it possible to have a certificate with multiple OEs or is that associate always a one-to-one relationship?  I appreciate that different OEs would require distinct test sessions.  From the API, I'm guessing that it's always a 1-to-1 relationship since the route to submit a test session for validation requires a single OE URL, but I wanted to confirm.

**Comment by smuellerDD** [comment]
Am Freitag, 6. März 2020, 19:59:00 CET schrieb Alex Thurston:

Hi Alex,

> This is more of a question than a bug.
> 
> It has to do with the organization of certificates/modules/oes.  I gather
> that it is possible (although it does not seem to be the norm) that a
> single module (or product as it's called on the CAVP list) can have
> multiple certificates.  Trolling through the current certificate list there
> are a handful of examples.  However, is it possible to have a certificate
> with multiple OEs or is that associate always a one-to-one relationship? 

See the Apple certs - there are very many OEs for one given module.

> I
> appreciate that different OEs would require distinct test sessions.

Yes, they do. But the module definition is identical during the publication 
phase. So, if the module definition is identical for different OEs, they all 
go into the same cert.

This even works retrospective. I.e. when you received one cert and sometimes 
later you, say, add a new OE or even add a new cipher, you extent the existing 
cert.

> From
> the API, I'm guessing that it's always a 1-to-1 relationship since the
> route to submit a test session for validation requires a single OE URL, but
> I wanted to confirm.



Ciao
Stephan

**Comment by AlexThurston** [comment]
Very interesting.   Thanks for that @smuellerDD .  I didn't realize that ACVP would consolidate if it discovered that the module already had a certificate.

If that's the case, then how does a situation like [this](https://csrc.nist.gov/projects/cryptographic-algorithm-validation-program/details?product=11750) occur?

In here, you have a module with two certs.  The OEs are different.  I thought it was because the algorithm capabilities were different but that they appear to be the same as well.  Shouldn't those two OEs be consolidated into a single certificate?

Also,
> This even works retrospective. I.e. when you received one cert and sometimes
> later you, say, add a new OE or even add a new cipher, you extent the existing
> cert.
If you add a new OE, thats makes sense since you would have a new test session but if you add a new algorithm capability, wouldn't that invalidate the pre-existing OEs since their test sessions at the time of submission would not have included that new algorithm?

**Comment by bfussell** [comment]
@AlexThurston I submitted some that ended up that way as well and the reason was that I submitted them all before the module had been manually approved so each one was interpreted as a new module rather than multiple OEs under the same module.  Lesson learned - submit the first one and get it manually approved before submitting additional OEs for the same module.

**Comment by smuellerDD** [comment]
Am Freitag, 6. März 2020, 20:32:46 CET schrieb Alex Thurston:

Hi Alex,

> Very interesting.   Thanks for that @smuellerDD .  I didn't realize that
> ACVP would consolidate if it discovered that the module already had a
> certificate.
> 
> If that's the case, then how does a situation like
> [this](https://csrc.nist.gov/projects/cryptographic-algorithm-validation-pr
> ogram/details?product=11750) occur?

That I do not know. But I expect that this consolidation does not work 
flawless yet on the server.

See also https://csrc.nist.gov/projects/cryptographic-algorithm-validation-program/details?product=12076

This Red Hat entry happend when the request was issued with one approval 
request. So, I consider that a glitch.

But I have added, say, the RSA 4k to A115 and A116 retrospectively (after the 
certs were issued). Or I added more OEs the Apple certs after the certs were 
issued.
> 
> In here, you have a module with two certs.  The OEs are different.  I
> thought it was because the algorithm capabilities were different but that
> they appear to be the same as well.  Shouldn't those two OEs be
> consolidated into a single certificate?
> 
> Also,
> 
> > This even works retrospective. I.e. when you received one cert and
> > sometimes later you, say, add a new OE or even add a new cipher, you
> > extent the existing cert.
> 
> If you add a new OE, thats makes sense since you would have a new test
> session

I do have a new test session.

> but if you add a new algorithm capability, wouldn't that invalidate
> the pre-existing OEs since their test sessions at the time of submission
> would not have included that new algorithm?

With the final PUT you tie the OE and module definitions together. So, if you 
have a different OE but the same module reference, the existing cert is 
extended.

Ciao
Stephan

**Comment by AlexThurston** [comment]
@bfussell That's a pretty compelling explanation.  I like it.

@smuellerDD For the second point, I don't think I was clear in what I was asking.  Lemme try and describe it - I apologize in advance for the long text.

I have a module `ACME` which registers a single algorithm `AES-ECB`.   It also has two OEs `Foo` and `Bar`.

I have two test sessions (IDs 1 and 2 for each OE respectively).  So I `PUT` for validation on test sessions 1 associating the module `ACME` with OE `Foo`.  This results in a certificate `A1`.

Then I PUT on test session 2 associating module `ACME` with OE `Bar`.  This - as you mentioned - would end up adding the OE `Bar` to the existing certificate `A1`.  Wonderful.  This is clear.

If I wanted to add more OEs, it would follow the exact same flow.  Great.

However, now I end up in a situation where I want to add a new cipher to an existing certificate.  Lets say its `AES-CBC`.  OK.  I make a new test session `3` that registers `AES-ECB` and `AES-CBC`, get responses from the vendor which they run using OE `Foo`.  Great.  Then I `PUT` on test session `3` for module `ACME` associating it with OE `Foo`.  Presumably that would have updated the cert to add `AES-CBC`.  Awesome

*But* - and here's what I was asking - wouldn't I also have to run another test session (`4`) through the vendors `Bar` OE as well.  Otherwise I'd have a situation where a different set of algorithms were run in each OE.

Hopefully this was too hard to follow.

**Comment by smuellerDD** [comment]
I had discussed this very issue with Shane some time back. It was clarified that there are sanity checks on the server that only merges new publications into existing certs when there is no inconsistency like you describe. If there would be an inconsistency, a new cert would be issued for the pending request.

Ciao
Stephan

—
Cui bono?

**Comment by AlexThurston** [comment]
Fair enough so when you said:
> This even works retrospective. I.e. when you received one cert and
> sometimes later you, say, add a new OE or even add a new cipher, you
> extent the existing cert.

Did you mean that that would work only if the certificate had a single OE?  If that were the case, would ACVP smarts understand that rather than create a new certificate, it would add the new cipher to the existing one?

**Comment by smuellerDD** [comment]
Am Freitag, 6. März 2020, 21:19:49 CET schrieb Alex Thurston:

Hi Alex,

> Fair enough so when you said:
> > This even works retrospective. I.e. when you received one cert and
> > sometimes later you, say, add a new OE or even add a new cipher, you
> > extent the existing cert.
> 
> Did you mean that that would work only if the certificate had a single OE? 

No. Look at the Apple certs. Each OE in there was added with an individual 
publication after the cert was issued before.

> If that were the case, would ACVP smarts understand that rather than create
> a new certificate, it would add the new cipher to the existing one?

If you publish the same set of ciphers for the same module but with a 
different OE, it gets added to the same cert, no matter how many OEs are 
already listed in the existing cert.

Ciao
Stephan

**Comment by AlexThurston** [comment]
What happens if you publish a test session with a different set of ciphers to the same module with the same OE?  And that module currently has a certificate with only has one OE?

**Comment by smuellerDD** [comment]
Am Freitag, 6. März 2020, 21:33:14 CET schrieb Alex Thurston:

Hi Alex,

> What happens if you publish a test session with a different set of ciphers
> to the same module with the same OE?

I never tried that, but I would expect that at least you will get rejected by 
the approval from NIST. I.e. you would need to get a new module definition.

> And that module currently has a
> certificate with only has one OE?



Ciao
Stephan

**Comment by AlexThurston** [comment]
@smuellerDD not to beat a dead horse here, but I'm still just a little confused on one last thing.  When you said

 > This even works retrospective. I.e. when you received one cert and
 > sometimes later you, say, add a new OE or even add a new cipher, you
 > extent the existing cert.

What did you mean by the second part _or even add a new cipher_?

Isn't that the scenario that I was describing?

**Comment by smuellerDD** [comment]
Am Montag, 9. März 2020, 14:09:51 CET schrieb Alex Thurston:

Hi Alex,

> @smuellerDD not to beat a dead horse here, but I'm still just a little
> confused on one last thing.  When you said
>  > This even works retrospective. I.e. when you received one cert and
>  > sometimes later you, say, add a new OE or even add a new cipher, you
>  > extent the existing cert.
> 
> What did you mean by the second part _or even add a new cipher_?
> 
> Isn't that the scenario that I was describing?

Add an OE:  you have a Cert with OE A and B and cipher definition Z

Now you can add to the same cert a test for OE C with the cipher definition Z


Add cipher definition: you have  a cert with OE A and cipher definition Z

Now you can add to the same cert a test for OE A and cipher definition Z plus 
another cipher.



Ciao
Stephan

**Comment by smuellerDD** [comment]
Can we close the issue?

**Comment by AlexThurston** [comment]
Yup.  I believe the original intent of this question/issue has been resolved.  Thanks again.
